### PR TITLE
P-333-evm-precompile

### PR DESCRIPTION
**Pull request by wangminqi**
P-333-evm-precompile

Adding evm precompile support for Rococo regarding parachain staking and token bridge with minimial feature.

Need more test case for both CI and parachain level test cases

### CODE REVIEW

File: precompiles/bridge-transfer/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	<Runtime as frame_system::Config>::AccountId,
+>>::Balance;
+
+/// A precompile to wrap the functionality from bridge_transfer.
+///
+/// EXAMPLE USECASE:
+/// A simple example usecase is a contract that allows donors to donate, and stakes all the funds
+/// toward one fixed address chosen by the deployer.
+/// Such a contract could be deployed by a collator candidate, and the deploy address distributed to
+/// supporters who want to donate toward a perpetual nomination fund.
```
Comment (Kailai-Wang) [inline precompiles/bridge-transfer/src/lib.rs]: Either delete or rewrite this section :)

File: precompiles/bridge-transfer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+{
+	fn transfer_native(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		input.expect_arguments(3)?;
+		let amount = input.read::<BalanceOf<Runtime>>()?;
+		let recipient: Vec<u8> = input.read::<Bytes>()?.into();
+		let dest_id: u8 = input.read::<u8>()?;
+
+		// Build call with origin.
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
```
Comment (Kailai-Wang) [inline precompiles/bridge-transfer/src/lib.rs:60]: Hmm, personally I'd still favor `HashedAddressMapping` over this one - does any other parachain use `HashedAddressMapping`?

Pull @kziemianek in too

File: precompiles/bridge-transfer/BridgeInterface.sol
Comment (Kailai-Wang) [inline precompiles/bridge-transfer/BridgeInterface.sol:1]: Have you tried to deploy the contract on the parachain? Just wondering if it's tested locally

File: precompiles/parachain-staking/src/lib.rs
Comment (Kailai-Wang) [inline precompiles/parachain-staking/src/lib.rs:1]: Does it make sense to further shrink the interface to only have absolutely needed calls?

**Review by Kailai-Wang — COMMENTED** [review]
Looks good just by browsing the code, but a lot comes down to the question if tests will pass.

The address conversion method needs to be further discussed tho

File: precompiles/parachain-staking/src/lib.rs
Comment (wangminqi) [reply precompiles/parachain-staking/src/lib.rs:1]: Sadly, they are already absolutely needed calls only.
Staking and unstaking, executing pending unstaking.

File: precompiles/parachain-staking/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,880 @@
...
+
+		// DelegationScheduledRequests:
+		// Blake2128(16) + AccountId(32)
+		// + Vec(
+		// 	ScheduledRequest(32 + 4 + DelegationAction(18))
+		// 	* (MaxTopDelegationsPerCandidate + MaxBottomDelegationsPerCandidate)
+		// )
+		handle.record_db_read::<Runtime>(
+			48 + (
+				56 * (<Runtime as pallet_parachain_staking::Config>::MaxTopDelegationsPerCandidate::get()
```
Comment (kziemianek) [inline precompiles/parachain-staking/src/lib.rs]: `32+4+18 = 54`
You put `56`, where `2` extra comes from ?

File: precompiles/utils/src/data.rs [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+
+// The implementation for u8 is specific, for performance reasons.
+impl EvmData for u8 {
+	fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
+		let range = reader.move_cursor(32)?;
+
+		let data = reader
+			.input
+			.get(range)
+			.ok_or_else(|| revert("tried to parse u64 out of bounds"))?;
```
Comment (kziemianek) [inline precompiles/utils/src/data.rs]: Should be `u8`

**Review by kziemianek — COMMENTED** [review]
Two small comments.
I'll continue reviewing tonight.

File: precompiles/bridge-transfer/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+		log::trace!(target: "bridge-transfer-precompile", "Execute input = {:?}", handle.input());
+
+		let selector = handle.read_selector()?;
+
+		handle.check_function_modifier(match selector {
+			Action::TransferNative => FunctionModifier::NonPayable,
+		})?;
+
+		match selector {
+			// read storage
```
Comment (kziemianek) [inline precompiles/bridge-transfer/src/lib.rs:97]: Maybe these comments can be removed ?

**Review by kziemianek — COMMENTED** [review]
In general it looks fine - I took a closer look at our code.
I left few small comments, only the one regarding gas usage looks important.

File: precompiles/bridge-transfer/src/lib.rs (hunk 1 above)
Comment (wangminqi) [reply precompiles/bridge-transfer/src/lib.rs]: Done

File: precompiles/bridge-transfer/src/lib.rs (hunk 2 above)
Comment (wangminqi) [reply precompiles/bridge-transfer/src/lib.rs:60]: I will make it in another PR for this hash method issue.

File: precompiles/parachain-staking/src/lib.rs (hunk 3 above)
Comment (wangminqi) [reply precompiles/parachain-staking/src/lib.rs]: you are right

File: precompiles/utils/src/data.rs (hunk 4 above)
Comment (wangminqi) [reply precompiles/utils/src/data.rs]: Done. YRR

File: precompiles/bridge-transfer/BridgeInterface.sol
Comment (wangminqi) [reply precompiles/bridge-transfer/BridgeInterface.sol:1]: Only test some of functions. Work well.

File: Makefile [hunk 6]
```diff
@@ -109,6 +109,10 @@ launch-binary-litmus:
 launch-binary-rococo:
 	@./scripts/launch-local-binary.sh rococo
 
+.PHONY: launch-custom-binary-rococo ## Launch a local rococo-parachain network with binary already built under target files
+launch-custom-binary-rococo:
+	@./scripts/launch-local-custom-binary.sh rococo
```
Comment (Kailai-Wang) [inline Makefile:114]: `launch-local-binary.sh` should be able to do the job already

**Review by Kailai-Wang — APPROVED** [review]
I approve it.

Can you please address the remaining comments and merge?

File: precompiles/bridge-transfer/src/lib.rs (hunk 5 above)
Comment (wangminqi) [reply precompiles/bridge-transfer/src/lib.rs:97]: I hope to save it for future format consistence.

File: Makefile (hunk 6 above)
Comment (wangminqi) [reply Makefile:114]: Please save it. Windows WSL make command will always rebuild bin locally if using launch-local-binary.sh ->->

**Pull request merged**